### ISSUE TITLE
[AMD][kernel][Tier 6] Port sol-execbench L2/061_sparse_moe_routing_and_experts to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#239]
# FlyDSL implementation: port sol-execbench's `061_sparse_moe_routing_and_experts` problem to FlyDSL

## Goal

Port the `061_sparse_moe_routing_and_experts` problem from `nvidia/SOL-ExecBench` (subset `L2`) to FlyDSL as part of the sol-execbench → FlyDSL coverage sweep. This is one of 235 problems across 4 subsets (L1, L2, Quant, FlashInfer-Bench) being tracked for FlyDSL kernel coverage.

**Source:** `nvidia/Qwen3-Next-80B-A3B-Thinking-NVFP4` (per `definition.json` `hf_id`)

**Description:** Sparse Mixture-of-Experts block with top-10 routing across 512 experts. Includes router network, top-k expert selection with normalized weights, parallel expert computation with weighted aggregation, and shared expert with learned gating. This is a complete MoE forward pass with dynamic token-to-expert routing.

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — kernel benchmark, not LLM workload)
Subset:    L2
Problem:   061_sparse_moe_routing_and_experts
Source:    nvidia/Qwen3-Next-80B-A3B-Thinking-NVFP4
```

**Const axes:**
  - `hidden_size` = 2048
  - `num_experts` = 512
  - `num_experts_per_tok` = 10
  - `moe_intermediate_size` = 512
  - `shared_expert_intermediate_size` = 512

**Inputs:**
  - `hidden_states` [batch_size, seq_len, hidden_size] bfloat16 — Input hidden states
  - `router_weight` [num_experts, hidden_size] bfloat16 — Router gate projection weight
  - `expert_gate_proj` [num_experts, moe_intermediate_size, hidden_size] bfloat16 — Expert gate projection weights (all experts stacked)
  - `expert_up_proj` [num_experts, moe_intermediate_size, hidden_size] bfloat16 — Expert up projection weights (all experts stacked)
  - `expert_down_proj` [num_experts, hidden_size, moe_intermediate_size] bfloat16 — Expert down projection weights (all experts stacked)
  - `shared_gate_proj` [shared_expert_intermediate_size, hidden_size] bfloat16 — Shared expert gate projection weight
  - `shared_up_proj` [shared_expert_intermediate_size, hidden_size] bfloat16 — Shared expert up projection weight
  - `shared_down_proj` [hidden_size, shared_expert_intermediate_size] bfloat16 — Shared expert down projection weight
  - `shared_expert_gate_weight` [1, hidden_size] bfloat16 — Shared expert gating weight

**Outputs:**
  - `output` [batch_size, seq_len, hidden_size] bfloat16 — Output hidden states after MoE processing
  - `router_logits` [batch_seq, num_experts] bfloat16 — Router logits for load balancing loss computation

**Workload samples (first 3):**
  - {"uuid": "6a24910a-e6e3-5311-bc57-d22f9fc4c1f7", "axes": {"batch_size": 2, "seq_len": 1664}, "inputs": {"hidden_states": {"type": "random"}, "router_weight": {"type": "random"}, "expert_gate_proj": {"
  - {"uuid": "0f72cc1c-6dfe-5ab4-a65f-a5fc090dc211", "axes": {"batch_size": 16, "seq_len": 224}, "inputs": {"hidden_states": {"type": "random"}, "router_weight": {"type": "random"}, "expert_gate_proj": {"
  - {"uuid": "fe300b07-fefb-58ff-bd36-d88a2b889de9", "axes": {"batch_size": 64, "seq_len": 128}, "inputs": {"hidden_states": {"type": "random"}, "router_weight": {"type": "random"}, "expert_gate_proj": {"

```text
Baseline:  sol-execbench PyTorch reference (reference.py) on AMD with torch.compile off
Target:    max_abs_diff < 1e-2 vs PyTorch reference on all tested shapes
           kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
```

## Scope

1. `kernels/sol_execbench/l2_061_sparse_moe_routing_and_experts.py` — FlyDSL kernel implementing the op described in `definition.json`, matching the PyTorch reference's numerics
2. `tests/kernels/test_sol_execbench_l2_061_sparse_moe_routing_and_experts.py` — parity harness vs PyTorch reference, testing multiple workload shapes
3. `tests/kernels/bench_sol_execbench_l2_061_sparse_moe_routing_and_experts.py` — reports `kernel_ms` per shape + comparison vs PyTorch eager baseline
4. `assessment.md` (≤300 words) — ranking which shapes FlyDSL wins/ties/loses vs PyTorch eager

## Acceptance

- Correctness: max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes
- Performance: kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
- Coverage: works on both gfx942 (MI300X) and gfx950 (MI355X)

## Out of scope

- Backward pass (unless the problem is itself a backward op)
- Weight gradients
- Optimizer step
- Cross-arch comparison vs NVIDIA H100 (PyTorch eager on AMD is the baseline)

## Reproduction

Clone `amdpilot-org/SOL-ExecBench`, download the dataset (`scripts/download_data.sh`), find the problem at `data/benchmark/L2/061_sparse_moe_routing_and_experts/`, run the sol-execbench driver against a FlyDSL solution. Requires the unified aiter+triton+flydsl Docker image at `/home/xiasun/unified-rocm.Dockerfile` (not yet built).

## Definition of done

- [ ] Kernel implemented at `kernels/sol_execbench/l2_061_sparse_moe_routing_and_experts.py`
- [ ] Tests pass on MI300X (gfx942)
- [ ] Tests pass on MI355X (gfx950)
- [ ] Bench shows ≥80% win rate vs PyTorch eager
- [ ] `assessment.md` written

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `061_sparse_moe_routing_and_experts` (kernel_id=155). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.462622 ms   (SOL-Score=1.0000, speedup=4.75×)
Scoring Baseline (optimized PyTorch): 2.115984 ms   (SOL-Score=0.5084, speedup=1.03×)
Reference Implementation (plain PyTorch): 73.709880 ms   (SOL-Score=0.0248, speedup=0.03×)

Top public submissions:
  - **Rank 1** (`doubleAI`): 0.748154 ms, SOL-Score=0.8557, speedup=2.94× vs reference
  - **Rank 2** (`Recursive`): 0.759898 ms, SOL-Score=0.8502
  - **Rank 3** (`Amir M. Mir | SF Tensor`): 0.947380 ms, SOL-Score=0.7804
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=2.94× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
